### DISCUSSION TITLE
Magazine Library Type

### DISCUSSION THREAD

**Discussion by DieselTech** [Kareadita/Kavita#2541] [Ideas]
### Idea Description

Posted by Nico
· Jan 2022
· 27 Upvotes
https://feats.kavitareader.com/posts/50/magazines-date-parsing

Description

I would like to add newspapers and magazines in my kavita library... I tried something but it's a mess. Could it be possible to create some kind of naming/parsing scheme similar to this for example ?
for a weekly/monthly magazine --> `//magazine/new-yorker/2022/new-yorker-2022-01-10.pdf`
and for daily papers = `//newspapers/jan2022/nyt-2022-01-10.pdf`

any other idea or suggestion are welcome ;-)

### Duration of Using Kavita

42

### Idea Category

Feature Enhancement

### Agreement

- [X] I agree that this is solely for submitting ideas, and I will search for existing ideas before posting.

**Comment by DieselTech** [comment]
themoot
· 12 months ago

Define a limited amount of valid formats that most magazines will adhere to. From my POV, these cover 95% of what I run into:

Date (ISO 8061) standard: YYYY-MM(-DD) (ie, 2023-01, 2023-01-19)
Date (Months): YYYY-Mon-[Mon] (ie, 2023-Jan-Feb; 2019-Jul-Aug-Sep)
Date (Season): YYYY-Spring/Summer/Fall/Winter (ie, 2023-Spring, 2021-Fall)
Vol/No = already handled?
Issue = already handled

Yes, you may have to normalize and rename files, but aren't we already doing that?

**Comment by majora2007** [comment]
This has been accepted. Kavita+ users can provide feedback to help me build out the initial version. I will be in contact with questions as necessary.

**Reply by majora2007** [reply]
If anyone has naming conventions and folder layouts, please share them here and I will try to build as much support as possible.

**Comment by majora2007** [comment]
Added a comment from https://feats.kavitareader.com/posts/450/minimal-interpretation-for-pdfs

I would also love to see some extra library types like "Magazine", "Article", "Leaflet", "Document", "Other", ... I think power users would appreciate a feature to specify one or multiple regular expressions that overwrite the default name parsing of Kavita and use the regex to get some metadata. e.g.

Microsoft - NET Microservices Architecture for Containerized NET Applications .NET 7.0 (2022).pdf
Insight Guides - South India (2017).epub
^(?<Publisher>.)\s-\s(?<Title>.)((?<ReleaseYear>[0-9]{4}))

National Geographic 2021-01.pdf
^(?<Title>.*)\s(?<ReleaseYear>[0-9]{4})-(?<ReleaseMonth>[0-9]{2})

Tolkien, J.R.R. - The Hobbit.epub
^(?<Writer>.)\s-\s(?<Title>.)

**Reply by DieselTech** [reply]
> "Magazine", "Article", "Leaflet", "Document", "Other"

Could this style of different options maybe be turned on per library like the advanced library options has the types selector?

![image](https://github.com/Kareadita/Kavita/assets/30128380/d123af72-d4be-4866-8bd5-c3dd0948efb0)

**Reply by majora2007** [reply]
I'm not sure what the differences between these are to be honest.

**Reply by DieselTech** [reply]
I would think the PDF parser would default to some standard, most likely Magazine. But articles (like research papers) might have a totally unique standard filename that they would want to overwrite the default PDF library with. I would think you would not want to rename research papers so that you can accurately find and reference the same one. Leaving them in their default name state might be needed.

**Reply by chiypg** [reply]
Would be nice to have a way to keep the default name. Not just research papers, but also manuals. And it may have type in file names. For example, HN678G4S6. Looks like the current parser doesn't handle this very well, and it generates wrong series names.

**Reply by dstibbe** [reply]
Is there not a way to create a custom parser? For regular books (PDF),  I like to group the books by collection (e.g. DnD 4e) without having to add 'DnD 4e' as a prefix to each file, or other dirty hacks (creating a dir for each pdf).
Just have a proper set of book rules instead of defaulting to 'manga rules'.  When I deal with books the most common case is that there are no separate files for chapters.

The current scanner is just a mess

**Reply by therobbiedavis** [reply]
> Is there not a way to create a custom parser? For regular books (PDF),  I like to group the books by collection (e.g. DnD 4e) without having to add 'DnD 4e' as a prefix to each file, or other dirty hacks (creating a dir for each pdf).
> 
> Just have a proper set of book rules instead of defaulting to 'manga rules'.  When I deal with books the most common case is that there are no separate files for chapters.
> 
> 
> 
> The current scanner is just a mess

This is an interesting idea, have an advanced option to let the admin write their own regex per pdf library.

**Reply by dstibbe** [reply]
Let's say someone wanted to edit the parser (for a PR), where would he start looking?

**Reply by majora2007** [reply]
Parser.cs but I would advise  you to speak to me before you get in too deep. Doing a POC is fine, but just because there is a PR, doesn't mean I'll pull it in. 

Can also look at the unit tests (MangaParserTests, etc)

**Reply by dstibbe** [reply]
Of course i don't expect anything to be 'just' accepted 😊

**Comment by majora2007** [comment]
For Magazines exclusively, the story to implement them is here:
https://github.com/Kareadita/Kavita/issues/2578

If you have common filenames, please comment on that. Do not comment about non-Magazine related information. This is exclusively designed around Magazines.

**Comment by huntsfromshadow** [comment]
You requested other organizational schemes.
This is an example of how I order a very large library of RPG pdfs.

Basically, I end up categorizing by publisher, then Game then book type then books.
Sometimes within a game, I'll have versions (Eg. Chaosium\Runequest\ then 3 folder there - RQ1 then RQ2, then RQG)

```
- RPG Books\
  - Masiff Publishing\
    - Lancer
      - Core Rules
        - Lancer Core.pdf
        - Character Sheet.pdf
      - Long Edge Setting
        - Setting Maps.pdf
        - Long Edge.pdf
  - Modiphius
    - Star Trek Adventures
      - Core Rules
        - STA.pdf
        - Utopia.pdf
      - Quick Start Box Set
        - Maps.pdf
        - Help Sheets.pdf
        - Rules.pdf
        - Setting.pdf
        - Campaign.pdf
      - Adventures
        - Lower Decks Adv
          - Lower Decks File.pdf
          - Lower Decks Map.pdf
          - Lower Decks Characters.pdf
```
  ```

**Comment by XyberDAWG** [comment]
> You requested other organizational schemes. This is an example of how I order a very large library of RPG pdfs.
> 
> Basically, I end up categorizing by publisher, then Game then book type then books. Sometimes within a game, I'll have versions (Eg. Chaosium\Runequest\ then 3 folder there - RQ1 then RQ2, then RQG)
> 
> ```
> - RPG Books\
>   - Masiff Publishing\
>     - Lancer
>       - Core Rules
>         - Lancer Core.pdf
>         - Character Sheet.pdf
>       - Long Edge Setting
>         - Setting Maps.pdf
>         - Long Edge.pdf
>   - Modiphius
>     - Star Trek Adventures
>       - Core Rules
>         - STA.pdf
>         - Utopia.pdf
>       - Quick Start Box Set
>         - Maps.pdf
>         - Help Sheets.pdf
>         - Rules.pdf
>         - Setting.pdf
>         - Campaign.pdf
>       - Adventures
>         - Lower Decks Adv
>           - Lower Decks File.pdf
>           - Lower Decks Map.pdf
>           - Lower Decks Characters.pdf
> ```
This is very similar to my Structure and Most likely the most common structure among TTRPG users. 

> ```
> Root
> - Publisher
> -- RPG (With Edition)
> --- Core Books
> --- Settings
> --- Adventures
> --- Supplements
> ```

So in these cases Each Folder below the "RPG" would in essence be treated as "Named Volumes"

**Comment by majora2007** [comment]
Just an update, I'm still thinking about a Magazine library, but with v0.8.5, PDF Metadata support (if tagged via Calibre) will be possible. Once this launches (soon), I would love some feedback on how much the magazine or other PDF-oriented (TTRPG) library types might be needed.

**Reply by XyberDAWG** [reply]
I do use it for a TTRPG library along side calibre. I am not sure if they are tagged via calibre or just side loaded. I haven't duginto calibre too deep. But I manage a TTRPG library separate in calibre.

**Comment by majora2007** [comment]
Just as an update, I swapped this from "Better PDF Parsing" into Magazine Support since that seemed to be the focus. With v0.8.5, Kavita can now parse PDF Metadata and thus allows for the original FR to be resolved. 

Please focus further comments around Magazine/Newspaper collections (or TTRPG) as a dedicated library option.

**Reply by majora2007** [reply]
Would anyone with Magazines be willing to help test a dedicated build and provide feedback? If so, please comment with your OS/Docker.

**Comment by TerryMathews** [comment]
I had commented on Reddit about using Kavita as a magazine reader (hi!) and asked to provide feedback here. I worked within the constraints of the system as it existed and utilized it's documentation to massage the files into a format where they would sort until I was happy with them.

I'm using version 0.8.6.2 (Docker, if that makes a difference).

Library type is Comic, I have disabled Epub, Manage Collections, Readling Lists, Scrobbing, Metadata Matching.

My folder structure is:

Magazines->National Lampoon->
National Lampoon - v{issue}.pdf

And then I also have one
SP1 High School Yearbook Parody.pdf

The result is 243 volumes in Volumes labled 1-243 and one in Specials labeled High School Yearbook Parody. It's not ideal, but it does work pretty well in a world where not all magazines follow regular issue cycles - National Lampoon gets to a point in it's run where it only issues twice a year before it dies.

I use this same structure for basically every magazine I have loaded in Kavita - Nintendo Power, Game Informer, the Sears Catalogs that I have, OXM, OMNI, EGM.

IMO, we need something like ID3 for PDFs to make this work right even if it lives externally. Maybe a JSON file - match the major part of the file name? We could spend a million man hours trying to come up with rules to cover every situation about how magazines number and we'd never cover them all. Better to just have a sidecar of the necessary data. In my humble opinion.

**Reply by majora2007** [reply]
Kavita does actually support metadata tagged from Calibre in PDF files. Since you have a lot of Magazines of different variations, what might an ideal Magazine parsing mechanism look like? Would different labeling be wanted for the card titles? 

Essentially, if you were to dream a bit (but realistically something that can fit into Kavita), what might it look like for your needs/reading patterns?

**Reply by JeffreyBytes** [reply]
Here's what ideal magazine parsing would look like from my perspective, based on the collections I've been trying to organize.

**The core problem: no universal numbering scheme**

I pulled a sample of titles I actively collect. Every one of them does something different:

- National Geographic - vol/number (`Volume 249 Number 06 June 2026`)
- Sky & Telescope - vol/number (`Volume 151 Number 3 March 2026`)
- FLYING - straight sequential (`Issue 971 June 2026`)
- Garden Rail - straight sequential (`Issue 383 July 2026`)
- WIRED - decimal encoding vol/issue (`Issue 34.02 March/April 2026`)
- The New Yorker - date only, sometimes a range (`May 11 & 18, 2026`)
- Flight Training - vol/number with combined bimonthly issues (`Volume 38 Number 02 March/April 2026`)
- Better Homes & Gardens Specials - title + year only, no issue number at all (`Cooking with Herbs, 2026`)

No filename parser will cover all of these correctly. WIRED's `34.02` looks like a float. The New Yorker has no issue number. BH&G Specials are essentially bookazines and have no number.

**What an ideal metadata model would look like**

Rather than trying to parse filenames, I think the right approach is a sidecar metadata file (JSON or a Kavita-specific XML) alongside each PDF. The fields that matter:

- `series` - magazine title
- `cover_date` - the one universal field across *every* magazine format (YYYY-MM or YYYY-MM-DD)
- `cover_date_display` - raw string for display purposes; handles ranges and irregular formats (`May 11 & 18, 2026` doesn't cleanly fit a date field)
- `issue_number` - sequential number where it exists (FLYING, Garden Rail)
- `volume` and `number` - for vol/issue formats (National Geographic, S&T)
- `combined_issue` - boolean flag for bimonthly/double issues
- `special_title` - for bookazines/specials that have no issue number

Three additional fields worth considering depending on scope:

- `issue_type` - controlled vocabulary (Regular, Special, Annual, Yearbook, Collector's Edition, etc.) rather than inferring type from whether `special_title` is populated.
- `edition`/`region` - some titles have meaningfully distinct regional editions that would otherwise collide on the same cover date. Technically solvable by including the region in the `series` field (e.g. `Better Homes & Gardens USA` vs `Better Homes & Gardens Australia`), but a dedicated field offers richer metadata and cleaner series grouping.
- `issn` - the canonical identifier for periodicals, analogous to ISBN for books. Note that ISSN identifies the series itself, not the individual issues, so this would be a series-level field rather than per-issue. 

**Cover date as the primary sort key**

Cover date is the only identifier that exists in every case - even The New Yorker has one. It's also how readers actually think about magazines ("the March issue," "the June 2026 issue"). I'd suggest making cover date the primary sort and display field, with issue number and vol/number as supplementary display fields shown on the card.

**Specials need a first-class bucket**

The current behavior of dumping specials into a "Specials" catch-all works, but it loses the title. BH&G Specials are meaningfully different from each other - `Cooking with Herbs` and `Summer Celebrations` aren't interchangeable the way numbered issues are. The special title should be the primary display label, not just overflow.

**What I'm not asking for**

I don't need Kavita to solve this through filename parsing rules. That rabbit hole is endless. I could add a dozen more edge cases from titles I didn't list here. A sidecar file that I populate (manually or via a tool) and Kavita reads is the cleaner solution. You already read Calibre-tagged PDF metadata, so the infrastructure for consuming external metadata is at least partially there.

Happy to provide more sample data or test anything if it helps.

**Comment by pred2k** [comment]
How can i get some kind of series grouping for magazines?

i've added my PDF collection which has the filesystem format `/<Magazine Name>/<Year>/<Magazine-Issue>/<File.pdf>` as library type=Image and after import all mag files are presented in a flat view. How can i get them grouped by `<Magazine Name>`?

Tagging all files could only be an option if mass-tagging with a extra data file would be possible.
